lib/rspec/matchers/built_in/be_instance_of.rb

```diff
@@ -5,16 +5,71 @@ module BuiltIn
       # Provides the implementation for `be_an_instance_of`.
       # Not intended to be instantiated directly.
       class BeAnInstanceOf < BaseMatcher
+        def initialize(*)
+          super
+          @no_checker_implemented = false
+        end
+
         # @api private
         # @return [String]
         def description
           "be an instance of #{expected}"
         end
 
+        # @api private
+        # @return [Boolean]
+        def matches?(actual)
+          @actual = actual
+          result = match(expected, actual)
+          @no_checker_implemented ? false : result
+        end
+
+        # @api private
+        # @return [Boolean]
+        def does_not_match?(actual)
+          @actual = actual
+          result = match(expected, actual)
+          @no_checker_implemented ? false : !result
+        end
+
+        # @api private
+        # @return [String]
+        def failure_message
+          if @no_checker_implemented
+            warning_message
+          else
+            super
+          end
+        end
+
+        # @api private
+        # @return [String]
+        def failure_message_when_negated
+          if @no_checker_implemented
+            warning_message
+          else
+            super
+          end
+        end
+
       private
 
         def match(expected, actual)
-          actual.instance_of? expected
+          if actual_object_respond_to?(actual, :instance_of?)
+            actual.instance_of?(expected)
+          else
+            @no_checker_implemented = true
```

Comment:
Rather than storing the result of the check, I suggest that you raise an `ArgumentError` here with the warning message and shortcut the rest.